armv7a9-zynq7000/syspage.c

```diff
@@ -0,0 +1,501 @@
+{
+	u8 id;
+	u32 newTop;
+
+	if (syspage_getMapID(map, &id) < 0) {
+		plostd_printf(ATTR_ERROR, "\nMAPS for %s doesn not exist!\n", map);
+		return -1;
+	}
+
+	newTop = (syspage_common.maps[id].top + PAGE_SIZE - 1) & ~(PAGE_SIZE - 1);
+
+	if (newTop > syspage_common.maps[id].map.end)
+		return -1;
+
+	syspage_common.maps[id].top = newTop;
+
+	return 0;
+}
+
+
+int syspage_getFreeSize(const char *map, u32 *sz)
+{
+	u8 id;
+
+	if (syspage_getMapID(map, &id) < 0) {
+		plostd_printf(ATTR_ERROR, "\nMAPS for %s doesn not exist!\n", map);
+		return -1;
+	}
+
+	*sz = syspage_common.maps[id].map.end - syspage_common.maps[id].top;
+
+	return 0;
+}
+
+
+int syspage_write2Map(const char *map, const u8 *buff, u32 len)
+{
+	u8 id;
+	u32 freesz;
+
+	if (syspage_getMapID(map, &id) < 0) {
+		plostd_printf(ATTR_ERROR, "\nMAPS for %s doesn not exist!\n", map);
+		return -1;
+	}
+
+	freesz = syspage_common.maps[id].map.end - syspage_common.maps[id].top;
+
+	if (freesz < len ){
+		plostd_printf(ATTR_ERROR, "\nThere isn't any free space in %s !\n", map);
+		return -1;
+	}
+
+	if (syspage_isMapFree(id, len) < 0 ) {
+		plostd_printf(ATTR_ERROR, "\nUps!! You encountered on some data. Top has been moved. Please try again!!\n");
+		return -1;
+	}
+
+	low_memcpy((void *)syspage_common.maps[id].top, buff, len);
+
+	syspage_common.maps[id].top += len;
+
+	return 0;
+}
+
+
+void syspage_addEntries(u32 start, u32 sz)
+{
+	int i;
+
+	for (i = 0; i < MAX_ENTRIES_NB; ++i) {
+		if (syspage_common.entries[i].start == 0 && syspage_common.entries[i].end == 0) {
+			syspage_common.entries[i].start = start;
+			syspage_common.entries[i].end = start + sz;
+			break;
+		}
+	}
+
+	/* Check whether entry overlapped with existing maps */
+	for (i = 0; i < syspage_common.mapsCnt; ++i)
+		syspage_addEntries2Map(i, start, start + sz);
+}
+
+
+
+/* Program's functions */
+
+int syspage_addProg(void *start, void *end, const char *imap, const char *dmap, const char *name)
+{
+	u8 imapID, dmapID;
+	unsigned int pos = 0, len = 0;
+	u32 progID = syspage_common.progsCnt;
+
+	if ((syspage_getMapID(imap, &imapID) < 0) || (syspage_getMapID(dmap, &dmapID) < 0)) {
+		plostd_printf(ATTR_ERROR, "\nMAPS for %s doesn not exist!\n", name);
+		return -1;
+	}
+
+	syspage_common.progs[progID].start = (u32)start;
+	syspage_common.progs[progID].end = (u32)end;
+	syspage_common.progs[progID].dmap = dmapID;
+	syspage_common.progs[progID].imap = imapID;
+
+	syspage_common.args[syspage_common.argCnt++] = 'X';
+	low_memcpy((void *)&syspage_common.args[syspage_common.argCnt], name, plostd_strlen(name));
+
+	syspage_common.argCnt += plostd_strlen(name);
+	syspage_common.args[syspage_common.argCnt++] = ' ';
+	syspage_common.args[syspage_common.argCnt] = '\0';
+
+	while (len < plostd_strlen(name)) {
+		if (name[len] == ';')
+			break;
+		++len;
+	}
+	low_memcpy(syspage_common.progs[progID].cmdline, name, len);
+
+	for (pos = len; pos < 16; ++pos)
+		syspage_common.progs[progID].cmdline[pos] = 0;
+
```

Comment:
```suggestion
	len = plostd_strlen(name);
	syspage_common.args[syspage_common.argCnt++] = 'X';
	low_memcpy((void *)&syspage_common.args[syspage_common.argCnt], name, len);

	syspage_common.argCnt += len;
	syspage_common.args[syspage_common.argCnt++] = ' ';
	syspage_common.args[syspage_common.argCnt] = '\0';

	while (pos < len) {
		if (name[pos] == ';')
			break;
		++pos;
	}
	low_memcpy(syspage_common.progs[progID].cmdline, name, pos);

	for (; pos < sizeof(syspage_common.progs[progID].cmdline); ++pos)
		syspage_common.progs[progID].cmdline[pos] = 0;

```
One `plostd_strlen()` instead  of three.
You can add `len` value checking for example like this:
```
	if (sizeof(syspage_common.args) - syspage_common.argCnt < (len + 3))
		return -1;

	if (sizeof(syspage_common.progs[progID].cmdline) < (len - 1))
		return -1;
```
Instead of `while{} low_memcpy() for{}` you can use one `for{}` .